---
title: "curatedTBData"
author: 
- name: Xutao Wang
  affiliation: 
  - Section of Computational Biomedicine, Boston University School of Medicine,
    Boston, MA, U.S.A.
  email: xutaow@bu.edu
package: curatedTBData
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{curatedTBData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Installtion
```{r install curatedTBData, eval = FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("curatedTBData")
devtools::install_github("compbiomed/curatedTBData")
```

# R packages
```{r libraries, message=FALSE, warning=FALSE, results='hide'}
library(curatedTBData)
library(dplyr)
library(SummarizedExperiment)
devtools::load_all()
library(curatedTBData)
```

# Access to MetaData
The `DataSummary` table summarized the list of available studies and their metadata information contained in the curatedTBData package. This helps users to query the datasets
in an easy and quick way.
```{r accessimg datasummary}
# Remove GeographicalRegion, Age, DiagnosisMethod, Notes, Tissue, HIVStatus for concise display
DataSummary %>%
  dplyr::select(-c(GeographicalRegion, Age, DiagnosisMethod, Notes,
                   Tissue, HIVStatus)) %>%
  DT::datatable()
```

# Access curated tuberculosis gene expression profile
Users can use `curatedTBData()` function to access data. There are three arguments in the function. The first argument `study_name` represents the names of the data that are used to determine the resources of interests. Users can find all available resource names from `DataSummary$GEOAccession`. The second argument `dryrun` enables users to determine the resources available before actually downloading them. When `dryrun` is set to `TRUE`, the returned values are titles of the resources. The third argument `curated.only` allows the users to access the curated ready-to-use data. If `curated.only` is `TRUE`, the function only download the curated gene expression profile and the clinical annotation of the corresponding data. When `curated.only` is `FALSE`, the function will download all the available resources of the corresponding data.
```{r access data with dryrun}
curatedTBData("GSE19439", dryrun = TRUE, curated.only = FALSE)
```
To download complete data for a Microarry study (e.g. [GSE19439](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19439)), 
we set `dryrun = FALSE` and `curated.only = FALSE`.
There are two experiments for the microarray study. The first experiment is `assay_curated`, which is a `matrix` that represents the normalized, curated version of the gene expression profile. The second experiment is `assay_raw`, which is a `r Biocpkg("SummarziedExperiment")` object that contains the raw gene expression profile and  information about probe features.
```{r download full data for GSE19439, warning=FALSE, results='hide'}
GSE19439 <- curatedTBData("GSE19439", dryrun = FALSE, curated.only = FALSE)
GSE19439
```
A third `experiment` called `assay_reprocess` is included, when accessing data for the RNA sequencing studies. This `matrix` represents the reprocessed version of gene expression profile from the raw fastq files using `r Biocpkg("Rsubread")`.
```{r, warning=FALSE, results='hide'}
GSE79362 <- curatedTBData("GSE79362", dryrun = FALSE, curated.only = FALSE)
GSE79362
```

A list of `r Biocpkg("MultiAssayExperiment")` objects is returned when `dryrun` is `FALSE`.
To save time, in the following example, we set the `curated.only = TRUE` and selected five studies that belong to [GSE19491](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19491) from the Gene Expression Omnibus. 
For RNA sequencing data, both curated assay and reprocessed assay plus the metadata information were returned.
```{r access selected curated datasets microarray, warning=FALSE, results='hide'}
myGEO <- c("GSE19435", "GSE19439", "GSE19442", "GSE19444", "GSE22098")
objectList <- curatedTBData(myGEO, dryrun = FALSE, curated.only = TRUE)
objectList[1:2]
```

```{r ACS RNA-seq, warning=FALSE, results='hide', eval=FALSE, include=FALSE}
# For RNA sequencing data, both curated assay and reprocessed assay plus the metadata information were returned.
# Here we retrieve the study from the adolescent cohort study [GSE79362](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE79362).
GSE79362 <- curatedTBData("GSE79362", dryrun = FALSE, curated.only = TRUE)
GSE79362
```

# Subset/Combine MultiAssayExperiment objects

## Subset
The major advantage of using `r Biocpkg("MultiAssayExperiment")` is the coordination of the meta-data and assays when sub-setting.The MultiAssayExperiment object has built-in function for subsetting samples based on column condition. For example, the following code shows how to select samples with only active TB.
```{r}
GSE19439 <- objectList$GSE19439
GSE19439[, GSE19439$TBStatus == "PTB"]@ExperimentList # 13 samples
```

The following code shows how to subset patients with active TB and LTBI
```{r}
GSE19439[, GSE19439$TBStatus %in% c("PTB","LTBI")]@ExperimentList # 30 samples
```

# Dataset integeration

## Merge Studies with common gene symbols
The `combineObjects()` function provides an easy implementation for merging different 
studies based on common gene symbols, it returned a `r Biocpkg("SummarizedExperiment")` object that contains 
the merged assay and associated clinical annotation. Noticed that [GSE74092](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi) is usually 
removed from merging, because this study used quantitative PCR so they did not have enough coverage to capture all genes. There two arguments in the `combineObjects()` function. The first one is `object_list`, which contains a list of `r Biocpkg("MultiAssayExperiment")` objects obtained from `curatedTBData()`. Notice that the `names(object_list)` should not be `NULL` and must be unique for each object within the list so that we can keep track the original study after merging. The second argument is `experiment_name`, which can be a string of vector of strings representing the name of the assay from the object. When the objects are merged, the original individual data tag can be found in the `Study` column by calling the  `colData()` on the output object.
```{r combine objects single experiment name}
GSE19491 <- combineObjects(objectList, experiment_name = "assay_curated")
GSE19491
# colData(GSE19491)$Study
```
It is also possible to combine the given list of objects with different experiment names. In this case, the `experiment_name` is vector of string that corresponds to the object within the input list.
```{r commbine objects multiple experiment name}
exp <- combineObjects(c(GSE79362[1], objectList[1]),
                      experiment_name = c("assay_reprocess_hg19", "assay_curated"))
exp
```
## Batch correction
If datasets are merged, it is typically recommended to remove a very likely batch effect. We will use the `ComBat()` method from `r Biocpkg("sva")` to remove potential batch effect between studies. 
In the following example, each study is viewed as one batch. The batch corrected
assay will be stored in the SummarizedExperiment object.
```{r batch correction, message=FALSE}
batch1 <- colData(GSE19491)$Study
combat_edata1 <- sva::ComBat(dat = assay(GSE19491), batch = batch1)
assays(GSE19491)[["Batch_corrected_assay"]] <- combat_edata1
GSE19491
```

# Example analysis - Active TB  vs. Latent infection
## Select samples with outecome of interest
The function `subset_curatedTBDatat()` allows the users to subset a list of `r Biocpkg("MultiAssayExperiment")` objects quickly. In the following example, we call this function to subset samples with active TB (`PTB`) and latent TB infection (`LTBI`) 
for binary classification. 
```{r subset active TB and LTBI, results = "hide", message=FALSE, warning=FALSE}
multi_set_PTB_LTBI <- lapply(objectList, function(x)
  subset_curatedTBData(x, annotationColName = "TBStatus", 
                       annotationCondition = c("LTBI","PTB"), 
                       assayName = "assay_curated"))
# Remove NULL from the list
multi_set_PTB_LTBI <- multi_set_PTB_LTBI[!sapply(multi_set_PTB_LTBI, is.null)]
```
## Profile multiple TB signatures across different studies
In the following example, we profile the performance of ten TB gene signatures from the `TBsignatures` using he `r Biocpkg("TBSignatureprofiler")` package, which evaluates the performance 
of a `list` of TB signatures using multiple gene set scoring methods. In the case, we selected ten candidates TB signatures from `TBSignatureProfiler::TBsignatures` and each subject's prediction score was computed using the single-sample Gene Set Enrichment Analysis scoring algorithm based on `r Biocpkg("GSVA")`. See `r Biocpkg("TBSignatureprofiler")` for more details about the `runTBsigProfiler()` function.
```{r, results = "hide", message=FALSE, warning=FALSE}
TBsignaturesSub <- TBSignatureProfiler::TBsignatures[1:10]
ssgsea_PTB_LTBI <- lapply(multi_set_PTB_LTBI,
                    function(x) TBSignatureProfiler::runTBsigProfiler(
                                  input = x,
                                  useAssay = assayNames(x)[1],
                                  signatures = TBsignaturesSub, 
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE))
```
## Visualization{.tabset}
### Area under the curve (AUC) and p-value results
The `combine_auc()` function allows the users to compute the area under the curve (AUC)
metrics for the prediction scores along with the bootstrapped confidence interval.
Notice that the `names()` of the input list should not be `NULL` and must be unique for each object within the list so that we can keep track the original study.
```{r, results='hide'}
# Obtain p.value, AUC, and Bootstrapped Confidence Interval
ssgsea_PTB_LTBI_combine <- combine_auc(ssgsea_PTB_LTBI, 
                                annotationColName = "TBStatus", 
                                signatureColNames = names(TBsignaturesSub), 
                                num.boot = 100, percent = 0.95)
DT::datatable(ssgsea_PTB_LTBI_combine)
```
### Mean AUC and bootstrapped confidence interval
The `bootstrap_mean_CI()` function also helps users compute the mean AUC value for each signature across studies and get the confidence interval at desired level.
```{r ssgsea summary table PTB vs. Latent}
# Select signatures and associated AUC, split them into list based on signature
ssgsea_PTB_LTBI_list <- ssgsea_PTB_LTBI_combine %>% 
  dplyr::select(.data$Signature, AUC) %>%
  dplyr::group_split(.data$Signature)

# Get summarized table and bootstrap 95% Confidence Interval for the mean AUC
sigInfo <- data.frame(Signature = sapply(ssgsea_PTB_LTBI_list, 
                                                     function(x) x$Signature[1]))
meanAUC <- do.call(rbind, lapply(ssgsea_PTB_LTBI_list, 
                                 function(x) bootstrap_mean_CI(x, colName = "AUC",
                                                               percent = 0.95, 
                                                               method = "empirical",
                                                               num.boot = 1000)))
ssgsea_PTB_LTBI_mean <- cbind(sigInfo, meanAUC)
DT::datatable(ssgsea_PTB_LTBI_mean)
```

### Boxplot of signature performance across studies
The `signatureBoxplot()` function from `r Biocpkg("TBSignatureProfiler")` provide great visualization of
multiple signature's performance on single study. Given a list of `SummarizedExperiment` object, the `BoxplotTBsig()` provides an alternative display such that it allows the comparison of signature's performance across different studies.
```{r ssgsea Boxplot for each signature PTB vs. LTBI, results = "asis", fig.width=3, fig.height=7}
for (i in names(TBsignaturesSub)) {
  cat("####", i, "\n")
  boxplotTBSig(ssgsea_PTB_LTBI, gset = i, annotationColName = "TBStatus")
  cat("\n\n")
}
```
### Ridge Plot
```{r}

```

### Heatmap

# Session Info
```{r session info}
sessionInfo()
```

