---
title: "curatedTBData"
author: 
- name: Xutao Wang
  affiliation: 
  - Section of Computational Biomedicine, Boston University School of Medicine,
    Boston, MA, U.S.A.
  email: xutaow@bu.edu
package: curatedTBData
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{curatedTBData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Installtion
```{r install curatedTBData, eval = FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("curatedTBData")
devtools::install_github("compbiomed/curatedTBData")
```

# R packages
```{r libraries, message=FALSE, warning=FALSE, results='hide'}
library(curatedTBData)
library(dplyr)
library(SummarizedExperiment)
# devtools::load_all()
library(curatedTBData)
```

# Access to MetaData
The `DataSummary` table summarized the list of available studies and their metadata information contained in the curatedTBData package. This helps users to query the datasets
in an easy and quick way.
```{r accessimg datasummary}
# Remove GeographicalRegion, Age, DiagnosisMethod, Notes, Tissue, HIVStatus for concise display
DataSummary %>%
  dplyr::select(-c(GeographicalRegion, Age, DiagnosisMethod, Notes,
                   Tissue, HIVStatus)) %>%
  DT::datatable()
```

# Access curated tuberculosis gene expression profile
Users can use `curatedTBData()` function to access data. There are three arguments in the function. The first argument `study_name` represents the names of the data that are used to determine the resources of interests. Users can find all available resource names from `DataSummary$GEOAccession`. The second argument `dryrun` enables users to determine the resources available before actually downloading them. When `dryrun` is set to `TRUE`, the returned values are titles of the resources. The third argument `curated.only` allows the users to access the curated ready-to-use data. If `curated.only` is `TRUE`, the function only download the curated gene expression profile and the clinical annotation of the corresponding data. When `curated.only` is `FALSE`, the function will download all the available resources of the corresponding data.
```{r access data with dryrun}
curatedTBData("GSE19439", dryrun = TRUE, curated.only = FALSE)
```
A list of `r Biocpkg("MultiAssayExperiment")` objects is returned when `dryrun` is `FALSE`.
To save time, in the following example, we set the `curated.only = TRUE` and selected five studies that belong to [GSE19491](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19491) from the Gene Expression Omnibus.
```{r access selected curated datasets microarray, warning=FALSE, results='hide'}
myGEO <- c("GSE19435", "GSE19439", "GSE19442", "GSE19444", "GSE22098")
objectList <- curatedTBData(myGEO, dryrun = FALSE, curated.only = TRUE)
objectList[1:2]
```

For RNA sequencing data, both curated assay and reprocessed assay plus the metadata information were returned.
Here we retrieve the study from the adolescent cohort study [GSE79362](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE79362).
```{r ACS RNA-seq, warning=FALSE, results='hide'}
GSE79362 <- curatedTBData("GSE79362", dryrun = FALSE, curated.only = TRUE)
GSE79362
```

# Subset/Combine MultiAssayExperiment objects

## Subset
The major advantage of using `r Biocpkg("MultiAssayExperiment")` is the coordination of the meta-data and assays when sub-setting.The MultiAssayExperiment object has built-in function for subsetting samples based on column condition. For example, the following code shows how to select samples with only active TB.
```{r}
GSE19439 <- objectList$GSE19439
GSE19439[, GSE19439$TBStatus == "PTB"]@ExperimentList # 13 samples
```

The following code shows how to subset patients with active TB and LTBI
```{r}
GSE19439[, GSE19439$TBStatus %in% c("PTB","LTBI")]@ExperimentList # 30 samples
```

# Dataset integeration

## Merge Studies with common gene symbols
The `combineObjects()` function provides an easy implementation for merging different 
studies based on common gene symbols, it returned a `r Biocpkg("SummarizedExperiment")` object that contains 
the merged assay and associated clinical annotation. Noticed that [GSE74092](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi) is usually 
removed from merging, because this study used quantitative PCR so they did not have enough coverage to capture all genes. There two arguments in the `combineObjects()` function. The first one is `object_list`, which contains a list of `r Biocpkg("MultiAssayExperiment")` objects obtained from `curatedTBData()`. Notices that the `names(object_list)` should not be `NULL` and must be unique for each object. The second argument is `experiment_name`, which can be a string of vector of strings representing the name of the assay from the object. When the objects are merged, the original individual data tag can be found in the `Study` column by calling the  `colData()` on the output object.
```{r combine objects single experiment name}
GSE19491 <- combineObjects(objectList, experiment_name = "assay_curated")
GSE19491

# colData(GSE19491)$Study
```
It is also possible to combine the given list of objects with different experiment names. 
In this case, the `experiment_name` is vector of string that corresponds to the object within the input list.
```{r commbine objects multiple experiment name}
exp <- combineObjects(c(GSE79362[1], objectList[1]),
                      experiment_name = c("assay_reprocess_hg19", "assay_curated"))
exp
```
## Batch correction
If datasets are merged, it is typically recommended to remove a very likely batch effect. We will use the `ComBat()` method from `r Biocpkg("sva")` to remove potential batch effect between studies. 
In the following example, each study is viewed as one batch:
```{r batch correction, message=FALSE}
batch1 <- colData(GSE19491)$Study
combat_edata1 <- sva::ComBat(dat = assay(GSE19491), batch = batch1)
assays(GSE19491)[["Batch_corrected_counts"]] <- combat_edata1
```

# Example analysis
##  Active TB  vs. Latent infection using `TBSignatureprofiler`
`r Biocpkg("TBSignatureprofiler")` is a package to evaluate the the performance of signatures across curated TB data. The function `subset_curatedTBDatat()` was used to subset samples with active TB (`PTB`) and latent TB infection (`LTBI`) for binary classification. In this example, each subject's prediction score was computed using the single-sample Gene Set Enrichment Analysis scoring algorithm from `r Biocpkg("GSVA")`. 
```{r calculate ssgsea for PTB vs LTBI, results = "hide", message=FALSE, warning=FALSE}
multi_set_PTB_Latent <- lapply(objectList, function(x)
  subset_curatedTBData(x, annotationColName = "TBStatus", 
                       annotationCondition = c("LTBI","PTB"), 
                       experiment_name = "assay_curated"))
```

# Session Info
```{r session info}
sessionInfo()
```

