---
title: "curatedTBData - Example"
author: "Xutao Wang"
output:
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
package: BiocStyle
vignette: >
  %\VignetteIndexEntry{curatedTBData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, cache.lazy = FALSE, message=FALSE, warning=FALSE)
```

```{r install curatedTBData, eval = FALSE}
devtools::install_github("compbiomed/curatedTBData")
```

```{r libraries, message=FALSE, warning=FALSE}
library(BiocParallel)
library(dplyr)
library(ggplot2)
library(ggridges)
library(gridExtra)
library(MultiAssayExperiment)
library(SummarizedExperiment)
library(TBSignatureProfiler)

devtools::load_all()
library(curatedTBData)
```

This package requires large memory to process data. Please increase R script memory before running this vignette. \
See the post for details: [Increase memory for Macs](https://stackoverflow.com/questions/51295402/r-on-macos-error-vector-memory-exhausted-limit-reached)\
Step 1: Open terminal\
Step 2:
```{r, engine='bash', eval=FALSE}
cd ~
touch .Renviron
open .Renviron
```
Step 3: Save the following as the first line of .Renviron:
```{r, engine='bash', eval=FALSE}
R_MAX_VSIZE=50Gb 
```

# `curatedTBData`: Clinically Annotated Data for the Tuberculosis (TB) Transcriptome

## Load available TB studies in the form of MultiAssayExperiemnt Objects

We use `MultiAssayExperiemnt` object to store information all available datasets. The pakcage function `curatedTBData` takes single or vector character string of Gene Expression Omnibus (GEO) accession as an input, and load the correpsonding curatedTBData.

```{r load a list of all datasets, results='hide'}
# To get a summary table for all the studies:
data("DataSummary", package = "curatedTBData")
DataSummary$`GEO accession`[-nrow(DataSummary)]
# Get a list of all data
# The following command takes about 5 miniutes
objects_list <- curatedTBData(DataSummary$`GEO accession`)
# Load single data
# curatedTBData("GSE39939")

# Load Multiple Studies
# curatedTBData(c("GSE39939","GSE39940"))
```

## Load all of the RNA-seq data
```{r load a list of RNA-seq studies, results='hide', eval=FALSE}
# To get a summary table for all the studies:
data("DataSummary", package="curatedTBData")
geo_RNA_seq <- DataSummary %>% 
  dplyr::filter(stringr::str_detect(.data$GeneralType, "RNA-seq")) %>%
  dplyr::select(.data$`GEO accession`)
object_list_RNA_seq <- curatedTBData(geo_RNA_seq$`GEO accession`)
```

## Load all of the Microarray data
```{r load a list of microarray studies, results='hide', eval=FALSE}
geo_microarray <- DataSummary %>% 
  dplyr::filter(stringr::str_detect(.data$GeneralType, "Microarray")) %>%
  dplyr::select(.data$`GEO accession`)

object_list_microarray <- curatedTBData(geo_RNA_seq$`GEO accession`)
```

# Subset Objects

The major adavantage of using `SummarizedExperiment` and `MultiAssayExperiment` is the coordination of the meta-data and assays when subsetting.
Both SummarizedExperiment and MultiAssayExperiment objects have similar buil-in fuction for subsetting samples based on column condition. For example, we want to select samples with only active TB, or with PTB or OD.
```{r subsetting single object example}
# subset samples from objects with TBStatus: PTB
GSE39939 <- objects_list$GSE39939 # 157 samples
GSE39939[, GSE39939$TBStatus == "PTB"] # 79 samples
GSE39939[, GSE39939$TBStatus %in% c("PTB","OD")] # 143 samples
```

# Datasets Integration

## Merge Studies with common gene symbols
If datasets are merged, it is typically recommended to remove a very likely batch effect. We will use the ComBat method, implemented for example in the SVA Bioconductor package. Start using `CombineObjects` to intersect objects with common gene symbol. This function produces a SummarizedExperiment object. GSE74092 is usually removed from merging, because this study used quantitative PCR so they did not have enough coverage to capture all genes.

```{r merge datasets, results="hide"}
# Merge seletced studies
GSE19491_geo <- c("GSE19435", "GSE19439", "GSE19442", "GSE19444", "GSE22098")
sobject <- CombineObjects(objects_list, 
                          list_name = GSE19491_geo, 
                          experiment_name = "assay_curated")
sobject
```

## Batch Correction
```{r batch correction, results='hide'}
# The samples that correspond to their datasets can be found at: colData(sobject)[,"GSE"]
# Batch correction, each study is viewed as one batch
batch1 <- colData(sobject)$GSE

combat_edata1 <- sva::ComBat(dat = as.matrix(assay(sobject)), batch = batch1)
SummarizedExperiment::assays(sobject)[["Batch_corrected_counts"]] <- combat_edata1

# To separate batch-corrected object into separate SummarizedExperiment object list:
batch_sobject_list <- lapply(unique(colData(sobject)[,"GSE"]), function(x) {
  index_GSE <- which(colData(sobject)$GSE %in% x)
  sobject[,index_GSE]
})
names(batch_sobject_list) <- unique(colData(sobject)[,"GSE"])


# Create a table for sample and data source information with:
sample_data_info <- data.frame(Sample = row.names(colData(sobject)), 
                               Data = colData(sobject)[,"GSE"])
head(sample_data_info)
```

# Possible outcome of interest could be patient's TB subtypes
The information about patient's TB type can be found at the column data with annotation name: TBStatus
```{r}
# Create a table for sample and TB status information with:
sample_outcome_info <- data.frame(Sample = row.names(colData(sobject)), 
                                  Data = colData(sobject)[,"TBStatus"])
head(sample_outcome_info)
```

# Active TB (PTB) vs. Latent TB (LTBI) using `TBSignatureprofiler`

## ssGSEA {.tabset}

`TBSignatureprofiler` is a package to evaluate the the performance of signatures across curated TB data. We subset samples with PTB and Latent TB status for binary classification using function `subset_curatedTBData`, and implement single-sample Gene Set Enrichment Analysis (ssGSEA) scoring algorithm to compute gene scores for each signature.
Workflow of ssGSEA:\
1. Transform sample-specific gene expression value into absolute rank. For gene $i$ and sample $j$, the rank is: $|r_{ij}|$. **The list is then ordered from the highest rank to the lowest**. \

2. After getting the ranked data, calculate random walk statistics for each sample. 
$$v_{jk}(\ell) = \frac{\sum_{i=1}^\ell |r_{ij}|^\tau I(g(i)\in\gamma_k)}{\sum_{i=1}^p|r_{ij}|^\tau I(g(i)\in\gamma_k)}-\frac{\sum_{i=1}^\ell I(g(i)\notin \gamma_k)}{p-|\gamma_k|}, \quad\ell=1,\dots,p\\
                ES_{jk} =\sum_{\ell=1}^p v_{jk}(\ell)$$
$\tau=0.25$.\
$p=$ Total number of genes.\
$|\gamma_k|$= Number of genes within gene set $k$.\

3. Re-scale the enrichment scores (ES) for each sample
$$ES_{jk}=\frac{ES_{jk}}{\max\{ES_{1k},\dots,ES_{nk}\}-\min\{ES_{1k},\dots,ES_{nk}\}}$$

### Calculate ssGSEA

```{r calculate ssgsea for PTB vs Latent, results = "hide", message=FALSE, warning=FALSE}
# Select sample contain information about PTB and Latent
# Modify the list to make it compatible with TBSignatureProfiler
param <- SerialParam(progressbar = TRUE)
multi_set_PTB_Latent <- bplapply(objects_list, function(x)
  subset_curatedTBData(x, annotationColName = "TBStatus", 
                       annotationCondition = c("LTBI","PTB"), 
                       experiment_name = "assay_curated"), 
                       BPPARAM = param) %>% plyr::compact()   
# Remove empty objects using plyr::compact()
TBsignatures <- TBsignatures[1:20]
ssgsea_PTB_Latent <- lapply(multi_set_PTB_Latent,
                    function(x) TBSignatureProfiler::runTBsigProfiler(
                                  input = x,
                                  useAssay = assayNames(x),
                                  signatures = TBsignatures, 
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE))

# Obtain p.value, AUC, and Bootstrapped Confidence Interval
ssgsea_PTB_Latent_combine <- combine_auc(ssgsea_PTB_Latent, 
                                annotationColName = "TBStatus", 
                                signatureColNames = names(TBsignatures), 
                                num.boot = 100, percent = 0.95)
```

### Boxplot of all scores signatures{.tabset}

```{r ssgsea Boxplot for each signature PTB vs. Latent, cache=TRUE, results = "asis", fig.width=30, fig.height = 32, fig.wide=TRUE}
for (i in names(TBsignatures)) {

  cat("####", i, "\n")
  
  BoxplotTBSig(ssgsea_PTB_Latent, gset = i, annotationColName = "TBStatus")
  
  cat("\n\n")
}
```


```{r include = FALSE}
# Need this chunk to ensure the inclusion/embedding of DT required javascript to work 
DT::datatable(NULL)
```

### Table with T-tests & AUC{.tabset}

```{r ssgsea Table for each signature PTB vs. Latent, cache=TRUE, results = "asis"}
for (i in names(TBsignatures)) {

  cat("####", i, "\n")
  
  print(htmltools::tagList(DT::datatable(ssgsea_PTB_Latent_combine %>%
                                           dplyr::filter(.data$Signature == i) %>%
                                           dplyr::select(-.data$Signature))))
  
  cat("\n\n")
}
```

### Summary Table for the performance of each signature

Compute mean AUC and bootstrapped 95% CI

```{r ssgsea summary table PTB vs. Latent}
# Seletct signatures and associated AUC
PTB_Latent_auc_summary_ssgsea <- ssgsea_PTB_Latent_combine %>% 
  dplyr::select(.data$Signature, AUC) %>%
  dplyr::group_split(.data$Signature)

# Get summarized table and bootstrap 95% Confidence Interval
ssgsea_PTB_Latent_auc_CI <- cbind(
  Signature=sapply(PTB_Latent_auc_summary_ssgsea, function(x) 
  x$Signature[1]), 
  do.call(rbind,lapply(PTB_Latent_auc_summary_ssgsea, function(x){
  bootstrap_mean_CI(x, colName="AUC", percent=0.95, num.boot=100)
}
)) 
)
ssgsea_PTB_Latent_auc_CI %>% DT::datatable()
```

### Comparison of Signature Performance using AUCs{.tabset}

#### Ridge Plot

```{r ssgsea AUC ridge plot for PTB vs Latent, fig.width=8, fig.height = 15, cache=TRUE}
# Ridge plot for AUCs distribution across datasets
# PTB vs. Latent
get_auc_distribution(ssgsea_PTB_Latent_combine) + 
  ggtitle("Ridge plot of AUC for PTB vs. Latent") + 
  theme(axis.text.x = element_text(colour="Black", size=12, hjust = 0.5, 
                                   face="bold"),
        axis.text.y = element_text(size=12, angle = 0, hjust = 0.5))

```

#### Heatmap

```{r ssgsea PTB vs. Latent heatmap with grouping, fig.width=10, fig.height = 15, fig.wide=TRUE}
# Import signature and data information
heatmap_auc(ssgsea_PTB_Latent_combine, SignatureInfoTraining, 
            names(TBsignatures), facet = TRUE) + 
  ggtitle("Heatmap of AUC for PTB vs. Latent")
```


## PLAGE {.tabset}

### Calculate ssGSEA

```{r plage calculate for PTB vs Latent, results = "hide"}
# Select sample contain information about PTB and Latent
# Modify the list to make it compatible with TBSignatureProfiler

plage_PTB_Latent <- lapply(multi_set_PTB_Latent,
                    function(x) TBSignatureProfiler::runTBsigProfiler(
                                  input = x,
                                  useAssay = assayNames(x),
                                  signatures = TBsignatures, 
                                  algorithm = "PLAGE",
                                  combineSigAndAlgorithm = TRUE))


# Obtain p.value, AUC
plage_PTB_Latent_combine <- combine_auc(plage_PTB_Latent, 
                                annotationColName = "TBStatus", 
                                signatureColNames = names(TBsignatures), 
                                num.boot = 100, percent = 0.95)
```

### Boxplot of all scores signatures{.tabset}

```{r plage Boxplot for each signature PTB vs. Latent, results = "asis", fig.width=30, fig.height = 32, fig.wide=TRUE, eval=FALSE}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  BoxplotTBSig(plage_PTB_Latent, gset = i, annotationColName = "TBStatus")
  
  cat("\n\n")
}
```


```{r include = FALSE}
# Need this chunk to ensure the inclusion/embedding of DT required javascript to work 
DT::datatable(NULL)
```

### Table with T-tests & AUC{.tabset}

```{r plage Table for each signature PTB vs. Latent, cache=TRUE, results = "asis", eval=FALSE}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  print(htmltools::tagList(DT::datatable(plage_PTB_Latent_combine %>%
                                           dplyr::filter(.data$Signature == i) %>%
                                           dplyr::select(-.data$Signature))))
  
  cat("\n\n")
}
```

### Summary Table for the performance of each signature

Compute mean AUC and bootstrapped 95% CI

```{r plage summary table PTB vs. Latent, eval=FALSE}
# Seletct signatures and associated AUC
PTB_Latent_auc_summary_plage <- plage_PTB_Latent_combine %>%
  dplyr::select(.data$Signature, AUC) %>%
  dplyr::group_split(.data$Signature)

# Get summarized table and bootstrap 95% Confidence Interval
plage_PTB_Latent_auc_CI <- cbind(
  Signature=sapply(PTB_Latent_auc_summary_plage, function(x) 
  x$Signature[1]), 
  do.call(rbind,lapply(PTB_Latent_auc_summary_plage, function(x){
  bootstrap_mean_CI(x, colName="AUC", percent=0.95, num.boot=100)
}
)) 
)
plage_PTB_Latent_auc_CI %>% DT::datatable()
```

### Comparison of Signature Performance using AUCs{.tabset}

#### Ridge Plot

```{r plage AUC ridge plot for PTB vs Latent, fig.width=8, fig.height = 15}
# Ridge plot for AUCs distribution across datasets
# PTB vs. Latent
get_auc_distribution(plage_PTB_Latent_combine) + 
  ggtitle("Ridge plot of AUC for PTB vs. Latent") + 
  theme(axis.text.x = element_text(colour="Black", size=12, hjust = 0.5, 
                                   face="bold"),
        axis.text.y = element_text(size=12, angle = 0, hjust = 0.5))

```
#### Ridge plot comparison between PLAGE and ssGSEA
```{r, fig.width=8, fig.height = 15}
plage_PTB_Latent_combine$Method <- "PLAGE"
ssgsea_PTB_Latent_combine$Method <- "ssGSEA"
d_all <- rbind(plage_PTB_Latent_combine, ssgsea_PTB_Latent_combine)
myColors <- RColorBrewer::brewer.pal(length(unique(d_all$Method)), "Set1")
ggplot(d_all,aes(x = AUC, y = Signature)) +
  geom_density_ridges(aes(fill = factor(Method)), alpha=0.8, scale=1.0) +
  scale_fill_manual(name = "Methods", values = myColors) + theme_bw()
```


#### Heatmap

```{r plage PTB vs. Latent heatmap with grouping, fig.width=10, fig.height = 15, fig.wide=TRUE}
# Import signature and data information
heatmap_auc(plage_PTB_Latent_combine, SignatureInfo, 
            names(TBsignatures), facet = TRUE) + 
  ggtitle("Heatmap of AUC for PTB vs. Latent (PLAGE)")
```

# Subset Patients with HIV
```{r PTB vs. Latent in HIV positive, results='hide', message=FALSE}
# Select objects with HIV information
object_match_HIV <- lapply(objects_list, function(x)
  check_annotation(x, annotationColName = "HIVStatus", 
                   annotationCondition = "Positive")) %>% plyr::compact()
length(object_match_HIV) # 7 studies with HIV positive patients
head(object_match_HIV)
```

# Select Patients with Diabetes Mellitus

```{r subset patients with Diabetes Mellitus, results='hide', message=FALSE}
# Select objects with Diabetes Positive
object_match_DA <- lapply(objects_list, function(x)
  check_annotation(x, annotationColName = "DiabetesStatus", 
                   annotationCondition = "Positive")) %>% plyr::compact()
object_match_DA
```
# Example on longitudinal datasets (i.e. subsetting samples at baseline)
There are 21 studies that have repeated measurement on the same patient, which could be used to deal with treatment response problems.
```{r}
# Studies involved with repeated measurement
# obtain studies with available patientID (not all NAs)
checkPatientIDName <- lapply(objects_list, function(x) {
  col_info <- colData(x)
  if(all(is.na(colData(x)$PatientID))) {
    isExist <- FALSE
  } else {
    isExist <- TRUE
  }
}) %>% unlist()
objects_list_long <- objects_list[checkPatientIDName]
length(objects_list_long)
```

We can also subset patients at baseline whenever it is possible. Next, we used the ACS cohort as an example.
```{r}
ACS_cohort <- objects_list$GSE79362
sample_baseline <- colData(ACS_cohort)[, c("PatientID","MeasurementTime")] %>%
  data.frame() %>%
  mutate(sample_name=row.names(colData(ACS_cohort))) %>%
  dplyr::arrange(MeasurementTime, PatientID) %>%
  dplyr::group_by(PatientID) %>%
  dplyr::mutate(first = dplyr::first(sample_name))
GSE79362_baseline <- ACS_cohort[,unique(sample_baseline$first)]
```

# Session Information
```{r session info}
sessionInfo()
```


